surefire-api/src/main/java/org/apache/maven/surefire/api/util/TempFileManager.java

```diff
@@ -0,0 +1,242 @@
+ * only in the system's temp directory.<br>
+ * This class creates temp files from a prefix, a unique date/timestamp, a short file id and suffix.
+ * It also helps you organize temporary files into sub-directories,
+ * and thus avoid bloating the temp directory root.<br>
+ * Apart from that, this class works in much the same way as {@link File#createTempFile(String, String)}
+ * and {@link File#deleteOnExit()}, and can be used as a drop-in replacement.
+ *
+ * @author Markus Spann
+ */
+public final class TempFileManager
+{
+
+    private static final Map<File, TempFileManager> INSTANCES  = new LinkedHashMap<>();
+    /** An id unique across all file managers used as part of the temporary file's base name. */
+    private static final AtomicInteger              FILE_ID    = new AtomicInteger( 1 );
+    private static final String                     SUFFIX_TMP = ".tmp";
+
+    private static Thread                           shutdownHook;
+
+    /** The temporary directory or sub-directory under which temporary files are created. */
+    private final File                              tempDir;
+    /** The fixed base name used between prefix and suffix of temporary files created by this class. */
+    private final String                            baseName;
+
+    /** List of successfully created temporary files. */
+    private final List<File>                        tempFiles;
+
+    /** Temporary files are delete on JVM exit if true. */
+    private boolean                                 deleteOnExit;
+
+    private TempFileManager( File tempDir )
+    {
+        this.tempDir = tempDir;
+        this.baseName = DateTimeFormatter.ofPattern( "yyyyMMddHHmmssSSS" ).format( LocalDateTime.now() );
+        this.tempFiles = new ArrayList<>();
+    }
+
+    private static File calcTempDir( String subDirName )
+    {
+        String tempDirName = subDirName == null ? null
+                        : subDirName.trim().isEmpty() ? null : subDirName.trim();
+        File tempDirCandidate = tempDirName == null ? new File( getJavaIoTmpDir() )
+                        : new File( getJavaIoTmpDir(), tempDirName );
+        return tempDirCandidate;
+    }
+
+    public static TempFileManager instance( )
+    {
+        return instance( ( File ) null );
+    }
+
+    /**
+     * Creates an instance using a subdirectory of the system's temporary directory.
+     * @param subDirName name of subdirectory
+     * @return instance
+     */
+    public static TempFileManager instance( String subDirName )
+    {
+        return instance( calcTempDir( subDirName ) );
+    }
+
+    public static synchronized TempFileManager instance( File tempDir )
+    {
+
+        // on creation of the first instance, register a shutdown hook to delete temporary files of all managers
+        if ( shutdownHook == null )
+        {
+            ThreadGroup tg = Thread.currentThread().getThreadGroup();
```

Comment:
Aren't those dead?

Reply:
Do you mean `ShutdownHook` or `ThreadGroup`?
Both are still part of moderns JDKs such as Java 17 (LTS) and neither marked deprecated nor discouraged.